gordo_components/builder/azure_utils.py

```diff
@@ -0,0 +1,470 @@
+
+    Example
+    -------
+    >>> s = json.dumps({"now":datetime.now(tz=UTC)}, cls=DatetimeEncoder, indent=4)
+    >>> s = '{"now": "2019-11-22 08:34:41.636356+"}'
+    """
+
+    def default(self, obj):
+        if isinstance(obj, datetime):
+            return obj.strftime("%Y-%m-%d %H:%M:%S.%f+%z")
+        else:
+            return json.JSONEncoder.default(self, obj)
+
+
+def send_artifacts(client: MlflowClient, run_id: str, metadata: dict):
+    """
+    Send artifacts
+
+    Parameters
+    ----------
+    client: MlflowClient
+        MLflow tracking client.
+    run_id: str
+        MLflow Run ID.
+    metadata: dict
+        Mode builder metadata to extract configs from.
+    """
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        for name, obj in [
+            ("model-config", metadata["model"]["model-config"]),
+            ("dataset-config", metadata["dataset"]),
+            ("user-config", metadata["user-defined"]),
+        ]:
+            fp = os.path.join(tmp_dir, f"{name}.json")
+            json.dump(obj, open(fp, "w"), cls=DatetimeEncoder)
+        client.log_artifacts(run_id, local_dir=tmp_dir)
+
+
+def get_mlflow_logger(
+    name: str,
+    model_key: str = uuid4().hex,
+    workspace_kwargs: dict = {},
+    service_principal_kwargs: dict = {},
+):
+    """
+    Generate MLflow logger function with either ia local or AzureML backend
+
+    Parameters
+    ----------
+    name: str
+        The name of the log group to log to (e.g. a model name).
+    model_key: str
+        Unique ID of logging run.
+    workspace_kwargs: dict
+        AzureML Workspace configuration to use for remote MLFlow tracking. See
+        :func:`gordo_components.builder.azure_utils.get_mlflow_client`.
+    service_principal_kwargs: dict
+        AzureML ServicePrincipalAuthentication keyword arguments. See
+        :func:`gordo_components.builder.azure_utils.get_mlflow_client`
+
+    Example
+    -------
+    >>> mlflow_logger = get_mlflow_logger("log_group", "unique_key", {}, {})
+    >>> mlflow_logger(metadata) # doctest: +SKIP
+    """
+
+    def mlflow_logger(metadata: dict):
+        """
+        Send logs to configured MLflow backend
+
+        Parameters
+        ----------
+        metadata: dict
+            Dictionary to log with MlflowClient
+        """
+
+        logger.info(
+            f"Logging dictionary with MLflow to {'AzureML' if workspace_kwargs else 'local backend'}"
+        )
+
+        mlflow_client = get_mlflow_client(workspace_kwargs, service_principal_kwargs)
+        run_id = get_run_id(mlflow_client, experiment_name=name, model_key=model_key)
+
+        # Log params and metrics
+        mlflow_client.log_batch(run_id, **get_batch_kwargs(metadata))
+
+        # Send configs as JSON artifacts
+        try:
+            send_artifacts(mlflow_client, run_id, metadata)
+        except Exception as e:
+            logger.error(e)
+
+        mlflow_client.set_terminated(run_id)
```

Comment:
This set_terminated looks like something that is a bit important to ensure gets called? Should it be in some finally block or something? In case an exception is thrown or something? What happens if this call fails, do we need to do some cleanup next time we log to this run_id?